future-apps/payroll/contracts/Payroll.sol

```diff
@@ -0,0 +1,569 @@
+     * @param initialYearlyDenominationSalary Employee's salary
+     */
+    function addEmployee(
+        address accountAddress,
+        uint256 initialYearlyDenominationSalary
+    )
+        public
+        auth(EMPLOYER_ROLE)
+    {
+        _addEmployee(
+            accountAddress,
+            initialYearlyDenominationSalary,
+            "",
+            getTimestamp()
+        );
+    }
+
+    /**
+     * @dev Add employee to Payroll
+     * @notice Add employee to Payroll. See addEmployeeWithNameAndStartDate
+     * @param accountAddress Employee's address to receive Payroll
+     * @param initialYearlyDenominationSalary Employee's salary
+     * @param name Employee's name
+     */
+    function addEmployeeWithName(
+        address accountAddress,
+        uint256 initialYearlyDenominationSalary,
+        string name
+    )
+        public
+        auth(EMPLOYER_ROLE)
+    {
+        _addEmployee(
+            accountAddress,
+            initialYearlyDenominationSalary,
+            name,
+            getTimestamp()
+        );
+    }
+
+    /**
+     * @dev Add employee to Payroll
+     * @notice Creates employee, adds it to mappings, initializes values.
+               Updates also global Payroll salary sum.
+     * @param accountAddress Employee's address to receive Payroll
+     * @param initialYearlyDenominationSalary Employee's salary
+     * @param name Employee's name
+     * @param startDate It will actually set initial lastPayroll value
+     */
+    function addEmployeeWithNameAndStartDate(
+        address accountAddress,
+        uint256 initialYearlyDenominationSalary,
+        string name,
+        uint256 startDate
+    )
+        public
+        auth(EMPLOYER_ROLE)
+    {
+        _addEmployee(
+            accountAddress,
+            initialYearlyDenominationSalary,
+            name,
+            startDate
+        );
+    }
+
+    /**
+     * @dev Set employee's annual salary
+     * @notice Updates also global Payroll salary sum
+     * @param employeeId Employee's identifier
+     * @param yearlyDenominationSalary Employee's new salary
+     */
+    function setEmployeeSalary(uint256 employeeId, uint256 yearlyDenominationSalary) public auth(EMPLOYER_ROLE) {
+        /* check that employee exists */
+        require(employeeIds[employees[employeeId].accountAddress] != 0);
+
+        yearlyTotalPayroll = yearlyTotalPayroll.sub(employees[employeeId].denominationTokenSalary.toYearlyDenomination());
+        employees[employeeId].denominationTokenSalary = yearlyDenominationSalary.toSecondDenominationToken();
+        yearlyTotalPayroll = yearlyTotalPayroll.add(employees[employeeId].denominationTokenSalary.toYearlyDenomination());
+    }
+
+    /**
+     * @dev Remove employee from Payroll
+     * @notice Updates also global Payroll salary sum
+     * @param employeeId Employee's identifier
+     */
+    function removeEmployee(uint256 employeeId) public auth(EMPLOYER_ROLE) {
```

Comment:
Lets make this a different role because `EMPLOYER_ROLE` will now have parameters specific to hiring. We can rename roles both also.